src/Compilers/Core/MSBuildTask/ManagedToolTask.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    public abstract class ManagedToolTask : ToolTask
+    {
+        protected abstract bool IsManagedTask { get; }
+
+        protected abstract string CommandLineArguments { get; }
+
+        protected abstract string PathToManagedTool { get; }
+
+        /// <summary>
+        /// Note: "Native" here does not neccesarily mean "native binary".
+        /// "Native" in this context means "native invocation", and running the executable directly.
+        /// </summary>
+        protected abstract string PathToNativeTool { get; }
+
+        protected sealed override string GenerateCommandLineCommands()
+        {
+            var commandLineArguments = CommandLineArguments;
+            if (IsManagedTask && IsCliHost(out string pathToDotnet))
+            {
+                var pathToTool = PathToManagedTool;
+                if (pathToTool is null)
+                {
+                    Log.LogErrorWithCodeFromResources("General_ToolFileNotFound", ToolName);
+                }
+                commandLineArguments = PrependFileToArgs(pathToTool, commandLineArguments);
+            }
+
+            return commandLineArguments;
+        }
+
+        protected sealed override string GenerateFullPathToTool()
+        {
+            if (IsManagedTask)
+            {
+                if (IsCliHost(out string pathToDotnet))
+                {
+                    return pathToDotnet;
+                }
+                else
+                {
+                    return PathToManagedTool;
+                }
+            }
+            else
+            {
+                return PathToNativeTool;
+            }
+        }
+
+        protected abstract string ToolNameWithoutExtension { get; }
+
+        protected sealed override string ToolName
+        {
+            get
+            {
+                if (CoreClrShim.IsRunningOnCoreClr)
```

Comment:
Should this condition include whether or not we have a managed tool? If the tool is not managed, it's never a DLL, right?

Reply:
No, because everything is terrible and on fire and also very very sad.

The comment at `ManagedCompiler.HasToolBeenOverridden` explains a bit of it - msbuild calls into these members really randomly, and one has to be extremely careful to not accidentally hit a recursion loop.

(I documented this a bit more, too)